Solution by robert

robertlisaru opened the pull request.

Commit:
dummy commit

Commit:
.gitignore

Commit:
my initial solutions + tests + test coverage html output

Commit:
some tweaks

Commit:
1st problem: neater solution in functional style. But now is tricky (for me) to do the input validation that checks for non-integers in the array (in a clean way, at least).

Commit:
I fixed the variable declarations.

Commit:
3rd problem in one line.

alexandru-calinoiu commented:
Please remove .idea from pr, I use Code to edit js so your .idea settings are of no value.

See ignore file https://github.com/github/gitignore/blob/9d7ff09c7d38dce9ef03e7ea4dc908a622546757/Global/JetBrains.gitignore

pirvudoru commented on solution/test/test-has-duplicates.js:
```diff
@@ -0,0 +1,19 @@
+    assert(hasDuplicates(numbers) == false);
+});
+
+//I'll skip the input validation tests for now
```
No need for this comment. Useful comments describe "WHY" not "WHAT"

robertlisaru replied:
I understand. Will clean-up. Thanks!

pirvudoru commented on solution/test/test-has-duplicates.js:
```diff
@@ -0,0 +1,19 @@
+    assert(hasDuplicates(numbers) == false);
+});
+
+it("test when duplicates", () => {
```
How will the following test case affect your solution?  ?
```
[1,1,4,4]
```

robertlisaru replied:
Ah, my solution is naive, should have thought more thoroughly. Will fix.

pirvudoru commented on solution/src/sum-evens.js:
```diff
@@ -0,0 +1,8 @@
+function sumEvens(numbers) {
+    if (!Array.isArray(numbers)) {
+        return undefined;
+    }
```
If you did not guard in `hasDuplicates` why guard here?

robertlisaru replied:
Just for the sake of exercising. But I'll assume input is valid (as stated in the problem description) and remove all the guards for a cleaner solution. Thank you!

pirvudoru commented on solution/test/test-has-duplicates.js:
```diff
@@ -0,0 +1,19 @@
+
+it("test duplicates in single number array", () => {
+    const numbers = [1];
+    assert(hasDuplicates(numbers) == false);
```
Why `assert(functionCall() == false)` and not just `assert(functionCall())` ?

robertlisaru replied:
Sorry, but is that correct? Shouldn't it be `assert( ! functionCall())` in order to be the same as `assert(functionCall() == false)` ?

pirvudoru commented on solution/coverage/largest-2.js.html:
```diff
@@ -0,0 +1,102 @@
+<span class="cline-any cline-yes">1x</span></td><td class="text"><pre class="prettyprint lang-js">function largest2(numbers) {
+    largest = [Number.MIN_VALUE, Number.MIN_VALUE];
+    for (i = 0; i &lt; numbers.length; i++) {
+        if (numbers[i] &gt; largest[0]) {
```
Too much noise with files that don't need to be here. Please clean-up so that your solution so it's more clear.

robertlisaru replied:
In the mail I got it says 

> Te rog sa incluzi un raport de test coverage in pr ca sa demonstrezi ca ai 100% coverage.

I wasn't sure what format should the report follow, so I included the _HTML_ report generated by _nyc_. 
Could you provide a few more details about how the report should look like?

pirvudoru commented on solution/test/test-sum-evens.js:
```diff
@@ -0,0 +1,26 @@
+it("sum evens when non integer values", () => {
+    const nonIntegers = [10, "definitely not an integer", 2, 3, 3.3];
+    assert.equal(sumEvens(nonIntegers), undefined);
+});
```
![image](https://user-images.githubusercontent.com/2707601/46124932-5f203d80-c22f-11e8-9d5c-96ae3661b566.png)

pirvudoru replied:
also the problem states:

Given an **integer array** of size N, compute the sum of all even numbers in this array

robertlisaru replied:
I understand. Will remove unnecessary tests.

pirvudoru commented on solution/test/test-largest-2.js:
```diff
@@ -0,0 +1,11 @@
+
+it("largest 2", () => {
+    const numbers = [1, 2, 3, 100, 4, 5, 6, 200];
+    let largest = largest2(numbers);
```
How will the following test case affect your solution? ?

```
[1, 2, 5, 4, 3]
```

robertlisaru replied:
Turns out my solution is wrong. Will fix.

pirvudoru commented on solution/package.json:
```diff
@@ -0,0 +1,16 @@
+    "mocha": "^5.2.0",
+    "nyc": "^13.0.1"
+  }
+}
```
How did you generate this file?

robertlisaru replied:
I run `npm init`, it asked me for most details.
Then I installed and saved _mocha_ test runner, `npm install --save-dev mocha`,
and _instanbul_ for test coverage reports `npm install --save-dev nyc` , which were added to "devDependencies".
Then I added a new script "test-with-coverage" which runs nyc and generates coverage report.
Is there a problem with the file?

pirvudoru commented on solution/src/sum-evens.js:
```diff
@@ -0,0 +1,8 @@
+    if (!Array.isArray(numbers)) {
+        return undefined;
+    }
+    return numbers.filter((x) => x % 2 == 0).reduce((x, y) => x + y, 0);
```
:100:

alexandru-calinoiu replied:
I will go with 90 not 100, it will be :100: if the variable names won't be some lame, `x` and `y`? Can you please be more expressive.

robertlisaru replied:
I've renamed to:
```javascript
return numbers.filter((number) => number % 2 == 0)
        .reduce((previousSum, currentValue) => previousSum + currentValue, 0);
```
Perhaps a bit better. Let me know if it's still not suggestive. Thank you!

pirvudoru reviewed (changes requested):
Nice first attempt but it requires more work. please see comments

pirvudoru commented:
Suggested reading: https://chris.beams.io/posts/git-commit/

Commit:
Update .gitignore for JetBrains files

Commit:
Remove instanbul generated html test coverage report

Commit:
Remove IDE specific file

Commit:
Add extra IDE file to .gitignore

Commit:
Remove unnecessary input validation test

Commit:
Remove input validation Unit Tests

Commit:
Add extra Unit Test scenarios

Also clean comments & rephrase somewhat redundant logic statements.

Commit:
Fix wrong solution to the largest 2 values problem

This solution is shorter, but has O(3n) complexity, iterating the array 3 times.

Commit:
Change the largest2 implementation to O(n) complexity

This implementation is longer, but only iterates one time through the array.

Commit:
Fix solution to duplicates problem

Commit:
Remove irellevant files

robertlisaru reviewed (commented):
This is the bit-vector approach, but I didn't explicitly allocate a bit-vector. The bit-vector is encoded in the input array by making the integers negative. So maybe I didn't break the rules?

Commit:
Clean .gitignore

alexandru-calinoiu commented on solution/src/has-duplicates.js:
```diff
@@ -0,0 +1,11 @@
+        if (numbers[Math.abs(numbers[i])] < 0) {
+            return true;
+        }
+        numbers[Math.abs(numbers[i])] *= -1;
```
This is a good solution.

You can make it easier to read by removing some duplication, no point on getting the absolute value two times.

robertlisaru replied:
That's right. Will fix.

alexandru-calinoiu commented on solution/src/largest-2.js:
```diff
@@ -0,0 +1,15 @@
+    let max = Number.MIN_VALUE;
+    let max2 = Number.MIN_VALUE;
+    for (let i = 0; i < numbers.length; i++) {
+        if (numbers[i] > max) {
```
Now that you have decent test coverage you can easily refactor this method, one thing you could to is remove duplication by removing the `number[i]` duplication. You've used `.filter` in the first problem can you also improve this problem?

robertlisaru replied:
I already have commited a solution using `.filter`
See commit: https://github.com/Agilefreaks/Test/pull/40/commits/0d0c716024591acea0d40ee52f827021b416ed44
It looks like this:
```javascript
function largest2(numbers) {
    let max = Math.max(...numbers);
    let max2 = Math.max(...numbers.filter((x) => x != max));
    return [max, max2];
}
```

alexandru-calinoiu commented on solution/test/test-has-duplicates.js:
```diff
@@ -0,0 +1,22 @@
+
+it("test duplicates in single number array", () => {
+    const numbers = [1];
+    assert(!hasDuplicates(numbers));
```
Negations are hard to reason about, can you write an assertion that will not require you to negate the result.

robertlisaru replied:
`assert(hasDuplicates(numbers) == false)` could be an alternative, but Doru says:
> Why `assert(functionCall() == false)` and not just `assert(functionCall())` ?

Another alternative:
`assert.equal(hasDuplicates(numbers), false)`

If something like "assertFalse" is what you're thinking of, I cannot find such a function:
https://nodejs.org/api/assert.html

alexandru-calinoiu commented on solution/test/test-largest-2.js:
```diff
@@ -0,0 +1,16 @@
+const largest2 = require('../src/largest-2');
```
you should have probably have a test with negative number

robertlisaru replied:
Right. Will include one. Thanks!

Commit:
Remove extra unnecessary calls to Math.abs in hasDuplicates

Commit:
Rename parameters for reduce function to be more expressive

Commit:
Add test case for largest2 when negative numbers

Commit:
Rename parameters for reduce function to be even more expressive

alexandru-calinoiu commented:
Will close this and followup via email.
